src/loader/mod.rs

```diff
@@ -0,0 +1,580 @@
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match self {
+            Error::BigEndianElfOnLittle => {
+                "Trying to load big-endian binary on little-endian machine"
+            }
+            Error::CommandLineCopy => "Failed writing command line to guest memory",
+            Error::CommandLineOverflow => "Command line overflowed guest memory",
+            Error::InvalidElfMagicNumber => "Invalid Elf magic number",
+            Error::InvalidProgramHeaderSize => "Invalid program header size",
+            Error::InvalidProgramHeaderOffset => "Invalid program header offset",
+            Error::InvalidProgramHeaderAddress => "Invalid Program Header Address",
+            Error::InvalidEntryAddress => "Invalid entry address",
+            Error::InvalidBzImage => "Invalid bzImage",
+            Error::InvalidKernelStartAddress => "Invalid kernel start address",
+            Error::InitrdImageSizeTooLarge => "Initrd image size too large",
+            Error::ReadElfHeader => "Unable to read elf header",
+            Error::ReadKernelImage => "Unable to read kernel image",
+            Error::ReadProgramHeader => "Unable to read program header",
+            Error::ReadBzImageHeader => "Unable to read bzImage header",
+            Error::ReadBzImageCompressedKernel => "Unable to read bzImage compressed kernel",
+            Error::ReadInitrdImage => "Unable to read initrd image",
+            Error::SeekKernelStart => "Unable to seek to kernel start",
+            Error::SeekElfStart => "Unable to seek to elf start",
+            Error::SeekProgramHeader => "Unable to seek to program header",
+            Error::SeekBzImageEnd => "Unable to seek bzImage end",
+            Error::SeekBzImageHeader => "Unable to seek bzImage header",
+            Error::SeekBzImageCompressedKernel => "Unable to seek bzImage compressed kernel",
+            Error::SeekInitrdImage => "Unable to seek initrd image",
+        }
+    }
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Kernel Loader Error: {}", Error::description(self))
+    }
+}
+
+/// * `kernel_load` - The actual `guest_mem` address where kernel image is loaded start.
+/// * `kernel_end` - The offset of `guest_mem` where kernel image load is loaded finish, return
+///                  in case of loading initrd adjacent to kernel image.
+/// * `setup_header` - The setup_header belongs to linux boot protocol, only for bzImage, vmm
+///                    will use it to setup setup_header.init_size, which is a must for bzImage
+///                    direct boot.
+#[derive(Debug, Default, Copy, Clone, PartialEq)]
+pub struct KernelLoaderResult {
+    pub kernel_load: GuestAddress,
+    pub kernel_end: GuestUsize,
+    pub setup_header: Option<bootparam::setup_header>,
+}
+
+pub trait KernelLoader {
+    fn load<F>(
+        guest_mem: &GuestMemoryMmap,
```

Comment:
Good question! VMM will pass mem to loader, so it's depends on VMM. If VMM use GuestMemoryMmap, but loader defined with GuestMemory, it will not work either. If VMM use GuestMemory, What GuestMemoryMmap is used for? I'm a bit confused. We need to have more discussions on this I think, event with other folks. I will just keep it as is at present.

Reply:
@bjzhjing are you joining the sync call today? I'll add this to the agenda.

Reply:
This should be generic indeed. Typically VMMs will pass mmap'ed memory to this call but there is no reason why this could not be generic and use `GuestMemory`.

Reply:
@sameo update with GuestMemory, but since cloud-hypervisor point to this crate, it requires changes there I think.

Reply:
@andreeaflorescu I update with GuestMemory, is it what you mean instead use traits (in slack)?

Reply:
Yes.

Reply:
@bjzhjing because of the `Cargo.lock` you won't break cloud-hypervisor. Go ahead and update this branch and then we can update in cloud-hypervisor after and bump the revision we use.

Reply:
@rbradford does this approach still work if you're doing push forces? The sha of the commit will be modified on push force. I noticed that you're using the same branch in cloud-hypervisor (master from @bjzhjing fork) as well. So unless you are using the cached repository to build cloud-hypervisor I think the build will fail.

Reply:
@andreeaflorescu yes, it will fail upon a force push, that is a good point.